homeassistant/components/tradfri/base_class.py

```diff
@@ -44,102 +35,27 @@ async def wrapper(command: Command | list[Command]) -> None:
-        self._device: Device = device
-        self._device_control: BlindControl | LightControl | SocketControl | SignalRepeaterControl | AirPurifierControl | None = (
-            None
-        )
-        self._device_data: Socket | Light | Blind | AirPurifier | None = None
-        self._gateway_id = gateway_id
-
-    async def _async_run_observe(self, cmd: Command) -> None:
-        """Run observe in a coroutine."""
-        try:
-            await self._api(cmd)
-        except PytradfriError as err:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed, trying again", exc_info=err)
-            self._async_start_observe()
-
-    @callback
-    def _async_start_observe(self, exc: Exception | None = None) -> None:
-        """Start observation of device."""
-        if exc:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed for %s", self._attr_name, exc_info=exc)
-        cmd = self._device.observe(
-            callback=self._observe_update,
-            err_callback=self._async_start_observe,
-            duration=0,
-        )
-        self.hass.async_create_task(self._async_run_observe(cmd))
-
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        self._async_start_observe()
-
-    @callback
-    def _observe_update(self, device: Device) -> None:
-        """Receive new state data for this device."""
-        self._refresh(device)
+        super().__init__(device_coordinator)
 
-    def _refresh(self, device: Device, write_ha: bool = True) -> None:
-        """Refresh the device data."""
-        self._device = device
-        self._attr_name = device.name
-        if write_ha:
-            self.async_write_ha_state()
-
-
-class TradfriBaseDevice(TradfriBaseClass):
-    """Base class for a TRADFRI device.
+        self._gateway_id = gateway_id
 
-    All devices should inherit from this class.
-    """
+        self._device: Device = device_coordinator.data
 
-    def __init__(
-        self,
-        device: Device,
-        api: Callable[[Command | list[Command]], Any],
-        gateway_id: str,
-    ) -> None:
-        """Initialize a device."""
-        self._attr_available = device.reachable
-        self._hub_available = True
-        super().__init__(device, api, gateway_id)
-
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        # Only devices shall receive SIGNAL_GW
-        self.async_on_remove(
-            async_dispatcher_connect(self.hass, SIGNAL_GW, self.set_hub_available)
-        )
-        await super().async_added_to_hass()
+        self._device_id = self._device.id
+        self._api = handle_error(api)
+        self._attr_name = self._device.name
 
-    @callback
-    def set_hub_available(self, available: bool) -> None:
```

Comment:
We should probably keep the gateway available callback in some form. Maybe move the callback to the coordinator instead and trigger the coordinator to create a new observation, eg by setting `last_update_success` to `False` and refreshing?

Reply:
It's not an option to set `self.available` to False in the coordinator's `_handle_exception` before it's refreshed? You added `and super().available` to the `TradfriBaseEntity`. Would it not catch this?
```
    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return cast(bool, self._device.reachable) and super().available

```

Reply:
Setting `self.last_update_success` to False in the coordinator will make the connected entities unavailable.

Reply:
The code I just added for `set_hub_available` runs when I spin up the dev server, disable wi-fi, execute a few commands, wait a while and then enable wi-fi again. What happens is that a connection is re-established to the gateway.